test/toolbox_image.sh

```diff
@@ -12,243 +12,65 @@ ARM_EMU_BIN="${ARM_EMU_BIN:-}"
-
-    if [ ! -d "${rootfs_dir}" ]; then
-        return
-    fi
-
-    for partition_table_file in "${rootfs_dir}${PARTITION_TABLE_FILE}"*; do
-        unlink "${partition_table_file}"
-    done
-
-    if [ -d "${rootfs_dir}${UPDATE_ROOTFS_SOURCE}" ]; then
-        rm -rf "${rootfs_dir}${UPDATE_ROOTFS_SOURCE}"
+    set -e
+    if [ -f "${toolbox_root_dir}/${TMP_TEST_IMAGE_FILE}" ]; then
+        unlink "${toolbox_root_dir}/${TMP_TEST_IMAGE_FILE}"
     fi
 
-    if [ -d "${update_mount}" ]; then
-        rm -r "${update_mount:?}"
-    fi
+    teardown_armhf_emulation "${toolbox_root_dir}"
 
-    if [ -b "${LOOP_STORAGE_DEVICE}" ]; then
-        losetup -d "${LOOP_STORAGE_DEVICE}"
+    if grep -qE "${toolbox_root_dir}" "/proc/mounts"; then
+        umount "${toolbox_root_dir}"
     fi
 
-    if [ -f "${rootfs_dir}${STORAGE_DEVICE_IMG}" ]; then
-        unlink "${rootfs_dir}${STORAGE_DEVICE_IMG}"
+    if [ -d "${toolbox_root_dir}" ]; then
+        rm -rf "${toolbox_root_dir:?}"
     fi
-
-    if [ -e "${rootfs_dir}/${ARM_EMU_BIN}" ]; then
-        if grep -q "$(realpath "${rootfs_dir}/${ARM_EMU_BIN}")" "/proc/mounts"; then
-            umount "${rootfs_dir}/${ARM_EMU_BIN}" || result=1
-        fi
-        if [ -f "${rootfs_dir}/${ARM_EMU_BIN}" ]; then
-            unlink "${rootfs_dir}/${ARM_EMU_BIN}" || result=1
-        fi
-    fi
-
-    if grep -q "${rootfs_dir}/dev" /proc/mounts; then
-        umount "${rootfs_dir}/dev" || result=1
-    fi
-
-    if "${is_dev_setup_mounted}"; then
-        umount "/dev" || result=1
-    fi
-
-    if grep -q "${rootfs_dir}/proc" /proc/mounts; then
-        umount "${rootfs_dir}/proc" || result=1
-    fi
-
-    for mount in ${mounts}; do
-        if grep -q "${mount}" /proc/mounts; then
-            umount "${mount}" || result=1
-        fi
-        if [ -d "${mount}" ]; then
-            rmdir "${mount}" || result=1
-        fi
-    done
 }
 
 failure_exit()
 {
     echo "Test exit per request."
     echo "When finished, the following is needed to cleanup!"
     echo "  sudo sh -c '\\"
-    echo "    umount '${rootfs_dir}/${ARM_EMU_BIN}' && \\"
-    echo "    unlink '${rootfs_dir}/${ARM_EMU_BIN}' && \\"
-    echo "    losetup -d '${LOOP_STORAGE_DEVICE}' && \\"
-    echo "    unlink '${STORAGE_DEVICE_IMG}' && \\"
-    echo "    umount '${rootfs_dir}/dev' && \\"
-    echo "    if '${is_dev_setup_mounted}'; then \\"
-    echo "      umount '/dev' \\"
-    echo "    fi && \\"
-    echo "    umount '${rootfs_dir}/proc' && \\"
-    echo "    umount '${rootfs_dir}' && \\"
-    echo "    rmdir '${rootfs_dir}' && \\"
-    echo "    umount '${overlayfs_dir}/rom' && \\"
-    echo "    rmdir '${overlayfs_dir}/rom' && \\"
-    echo "    umount '${overlayfs_dir}/' && \\"
-    echo "    rmdir '${overlayfs_dir}/'"
+    echo "    unlink '${toolbox_root_dir}/${TMP_TEST_IMAGE_FILE}'&& \\"
+    echo "    umount '${toolbox_root_dir}' && \\"
+    echo "    rm -rf '${toolbox_root_dir:?}' && \\"
```

Comment:
I'm not sure if the ? is wise here

if toolbox_root_dir is NULL (the -u should protect you for the unset variable at the very start of this function/script) it will print:
unlink 'NULL'/tmp_test_image.img && \
umount 'NULL' && \
<exit 1>